extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java

```diff
@@ -0,0 +1,307 @@
+import org.apache.druid.frame.channel.ReadableFrameChannel;
+import org.apache.druid.frame.channel.WritableFrameChannel;
+import org.apache.druid.frame.key.ClusterBy;
+import org.apache.druid.frame.processor.FrameProcessor;
+import org.apache.druid.frame.processor.OutputChannel;
+import org.apache.druid.frame.processor.OutputChannelFactory;
+import org.apache.druid.frame.processor.OutputChannels;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.Pair;
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.msq.counters.CounterTracker;
+import org.apache.druid.msq.input.InputSlice;
+import org.apache.druid.msq.input.InputSliceReader;
+import org.apache.druid.msq.input.InputSlices;
+import org.apache.druid.msq.input.ReadableInput;
+import org.apache.druid.msq.input.ReadableInputs;
+import org.apache.druid.msq.input.external.ExternalInputSlice;
+import org.apache.druid.msq.input.stage.StageInputSlice;
+import org.apache.druid.msq.kernel.FrameContext;
+import org.apache.druid.msq.kernel.ProcessorsAndChannels;
+import org.apache.druid.msq.kernel.StageDefinition;
+import org.apache.druid.segment.column.RowSignature;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+public abstract class BaseLeafFrameProcessorFactory extends BaseFrameProcessorFactory
+{
+  private static final Logger log = new Logger(BaseLeafFrameProcessorFactory.class);
+
+  @Override
+  public ProcessorsAndChannels<FrameProcessor<Long>, Long> makeProcessors(
+      StageDefinition stageDefinition,
+      int workerNumber,
+      List<InputSlice> inputSlices,
+      InputSliceReader inputSliceReader,
+      @Nullable Object extra,
+      OutputChannelFactory outputChannelFactory,
+      FrameContext frameContext,
+      int maxOutstandingProcessors,
+      CounterTracker counters,
+      Consumer<Throwable> warningPublisher
+  ) throws IOException
+  {
+    // BaseLeafFrameProcessorFactory is used for native Druid queries, where the following input cases can happen:
+    //   1) Union datasources: N nonbroadcast inputs, which are are treated as one big input
+    //   2) Join datasources: one nonbroadcast input, N broadcast inputs
+    //   3) All other datasources: single input
+
+    final int totalProcessors = InputSlices.getNumNonBroadcastReadableInputs(
+        inputSlices,
+        inputSliceReader,
+        stageDefinition.getBroadcastInputNumbers()
+    );
+
+    if (totalProcessors == 0) {
+      return new ProcessorsAndChannels<>(Sequences.empty(), OutputChannels.none());
+    }
+
+    final int outstandingProcessors;
+
+    if (hasParquet(inputSlices)) {
```

Comment:
A couple thoughts:

1. We could provide a way for InputFormats to declare how much memory they expect to use, and size the number of concurrent processors accordingly.
2. AFAIK, this is only a problem with Parquet. We could use a different Parquet reader, or write our own, such that this extra memory use doesn't happen. For example we could memory-map the Parquet files, instead of loading them up into the heap as the current reader does.